Add Document support to _WKSerializedNode

achristensen07 opened the pull request:
#### 00d5122b17f1d0baec9835d7854ce49b7714d0a3
<pre>
Add Document support to _WKSerializedNode
<a href="https://bugs.webkit.org/show_bug.cgi?id=296506">https://bugs.webkit.org/show_bug.cgi?id=296506</a>
<a href="https://rdar.apple.com/156746228">rdar://156746228</a>

Reviewed by Ryosuke Niwa.

This PR is mostly moving things around to be shared between node cloning and serialization.
I begin to add the shape of cloning ContainerNodes, but it&apos;s not really implemented much yet.
I add IDL support for serializeNode(node, {deep:true}).
I changed cloneDocumentWithoutChildren from a virtual function to Document::clonedDocumentType
and Document::createCloned which know about all supported subclasses of Document.
When deserializing a Document, some properties are taken from the serialized values and some
are taken from the context Document into which you are deserializing a Node.
JavaScriptEvaluationResult::Value has become too big to fit into a HashMap value without
hitting the static assertion in HashMap::inlineLookup, so I switch to a UniqueRef&lt;Value&gt;.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::cloneChildNodes const):
(WebCore::ContainerNode::serializeChildNodes const):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clonedDocumentType const):
(WebCore::Document::cloneNodeInternal const):
(WebCore::Document::serializeNode const):
(WebCore::Document::createCloned):
(WebCore::Document::cloneDocumentWithoutChildren const): Deleted.
(WebCore::Document::cloneDataFromDocument): Deleted.
* Source/WebCore/dom/Document.h:
(WebCore::Document::compatibilityMode const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::deserializeNode): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/SerializedNode.cpp: Added.
(WebCore::SerializedNode::deserialize):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/html/HTMLDocument.cpp:
(WebCore::HTMLDocument::cloneDocumentWithoutChildren const): Deleted.
* Source/WebCore/html/HTMLDocument.h:
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::serializeNode):
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebCore/page/WebKitSerializedNode.cpp:
(WebCore::WebKitSerializedNode::WebKitSerializedNode):
* Source/WebCore/page/WebKitSerializedNode.h:
(WebCore::WebKitSerializedNode::create):
* Source/WebCore/svg/SVGDocument.cpp:
(WebCore::SVGDocument::cloneDocumentWithoutChildren const): Deleted.
* Source/WebCore/svg/SVGDocument.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::toString const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::map const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toValue):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: <a href="https://commits.webkit.org/297895@main">https://commits.webkit.org/297895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29eb985eea95608aea2eb2d44780cdc790b26408

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113291 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33031 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23504 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119499 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63999 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115253 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41589 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86225 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116238 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26905 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101911 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66552 "Found 141 new API test failures: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration, /WPE/TestBackForwardList:/webkit/WebKitWebView/navigation-after-session-restore, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/reset, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WPE/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone-worker, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/injected-script, /WPE/TestLoaderClient:/webkit/WebKitWebView/unfinished-subresource-load, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/cookies ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26170 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20037 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63253 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96281 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/20113 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122716 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30117 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98123 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94817 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39965 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17782 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36515 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18205 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40255 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45754 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39896 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43229 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5470,10 +5471,23 @@ bool Document::canAcceptChild(const Node& newChild, const Node* refChild, Accept
+    if (isXMLDocument()) {
+        if (isXHTMLDocument())
+            return ClonedDocumentType::XMLDocument;
+        return ClonedDocumentType::XMLDocument;
```
If these both return XMLDocument, why have the branch for isXHTMLDocument? (Edit: it seems like this is a bug, as below we create a type for XHTML as well.)

achristensen07 replied:
XHTML was an oversight.  Thanks!

annevk commented on Source/WebCore/dom/SerializedNode.h:
```diff
@@ -25,35 +25,54 @@
+        URL baseURL;
+        URL baseURLOverride;
+        Variant<String, URL> documentURI;
+        String contentType;
```
FWIW, the specification also lists encoding and allow declarative shadow roots as aspects to be copied. (And origin and custom element registry, but I don't think those are applicable here.)

achristensen07 replied:
I don't think we've implemented "allow declarative shadow roots" yet.  I arbitrarily decided to inherit the encoding from the deserialization document, but may revisit.

achristensen07 committed:
Add Document support to _WKSerializedNode
https://bugs.webkit.org/show_bug.cgi?id=296506
rdar://156746228

Reviewed by Ryosuke Niwa.

This PR is mostly moving things around to be shared between node cloning and serialization.
I begin to add the shape of cloning ContainerNodes, but it's not really implemented much yet.
I add IDL support for serializeNode(node, {deep:true}).
I changed cloneDocumentWithoutChildren from a virtual function to Document::clonedDocumentType
and Document::createCloned which know about all supported subclasses of Document.
When deserializing a Document, some properties are taken from the serialized values and some
are taken from the context Document into which you are deserializing a Node.
JavaScriptEvaluationResult::Value has become too big to fit into a HashMap value without
hitting the static assertion in HashMap::inlineLookup, so I switch to a UniqueRef<Value>.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::cloneChildNodes const):
(WebCore::ContainerNode::serializeChildNodes const):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clonedDocumentType const):
(WebCore::Document::cloneNodeInternal const):
(WebCore::Document::serializeNode const):
(WebCore::Document::createCloned):
(WebCore::Document::cloneDocumentWithoutChildren const): Deleted.
(WebCore::Document::cloneDataFromDocument): Deleted.
* Source/WebCore/dom/Document.h:
(WebCore::Document::compatibilityMode const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::deserializeNode): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/SerializedNode.cpp: Added.
(WebCore::SerializedNode::deserialize):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/html/HTMLDocument.cpp:
(WebCore::HTMLDocument::cloneDocumentWithoutChildren const): Deleted.
* Source/WebCore/html/HTMLDocument.h:
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::serializeNode):
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebCore/page/WebKitSerializedNode.cpp:
(WebCore::WebKitSerializedNode::WebKitSerializedNode):
* Source/WebCore/page/WebKitSerializedNode.h:
(WebCore::WebKitSerializedNode::create):
* Source/WebCore/svg/SVGDocument.cpp:
(WebCore::SVGDocument::cloneDocumentWithoutChildren const): Deleted.
* Source/WebCore/svg/SVGDocument.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::toString const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::map const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toValue):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: https://commits.webkit.org/297895@main
